src/bootstrap/tool.rs

```diff
@@ -273,6 +273,10 @@ pub fn prepare_tool_cargo(
     if !features.is_empty() {
         cargo.arg("--features").arg(&features.join(", "));
     }
+
+    // librustc_driver is linked as both an rlib and a dylib. Prefer the dylib.
+    cargo.rustflag("-Cprefer-dynamic");
```

Comment:
This means that everyone using `rustc_driver` will need to pass this flag themself, or get a confusing error message.

Reply:
Yes, I think the tests pass now, but I don't actually think this is really a good idea. I still don't understand the error message, btw.

I'll try adding a separate librustc_driver_static instead, so the existing interface still works.

Reply:
Here, I put it in a separate branch: https://github.com/rust-lang/rust/compare/master...sapir:rustc-driver-rlib-extra